### ISSUE TITLE
Limit submission rate of jobs

### ISSUE THREAD

**Issue by murphycj** [nextflow-io/nextflow#198]
The number of concurrent processes that can be run is already a feature. This issue proposes a feature that would allow the rate at which processes are submitted to a scheduler (e.g. SGE) to reduce the likelihood of over-taxing the file system.

For example, if I have 100 jobs then allow me to specify an option to submit 10 jobs, wait 2 minutes, submit 10 jobs, wait another 2 minutes, etc.. The countdown timer between submissions should start after the previous set of jobs have started running on their job nodes. This issue is in reference to the problem raised [here](https://groups.google.com/forum/#!topic/nextflow/YtxVVLa0L7g).

**Comment by murphycj** [comment]
Putting this here for myself too, but if anyone has the same issue you can put the following at the beginning of the command execution block of the process to randomly wait between 1-500 seconds before executing

    if hash shuf >/dev/null; then
        sleep $( shuf -i 1-500 -n 1 )
    fi

**Comment by dctrud** [comment]
Alternatively:

`sleep $[ ( $RANDOM % 10 )  + 1 ]s`

where `10` is the max wait might be more portable, as `shuf` is not available on OSX by default.

**Comment by pditommaso** [comment]
More in general why limiting the submission rate should be better to limiting the max number of jobs queued by the cluster?

**Comment by murphycj** [comment]
It could be better if each individual job takes a long time (e.g. several days), which I have encountered with my RNA-seq and whole-exome seq datasets. Say I' am doing an RNA-seq analysis that takes 2 days/sample. Then thats just 2 days of waiting if I submit all 100 samples or 10 days if I submit in batches of 20.

Admittedly, I think my case is probably not really that common, so maybe this feature would not get used that often. But submitting 100 RNA-seq jobs, each needing to copy over gigabytes of data to the job server, does tax the file system and affect other users, but its only temporary (e.g. 10-20 minutes). But the fixes stated above by dctrud and I are super easy to implement. So up to you if you want to implement this feature.

**Comment by pditommaso** [comment]
Yes, but what I'm saying that you can set the max number of jobs that can be submitted in parallel, let's say to 10 instead of 10. It would not be a workaround to avoid to overload the cluster/file system?

**Comment by murphycj** [comment]
I' am not sure I understand you. My cluster system (SGE) only allows me to limit the number of jobs that are being run in parallel (as far as I know), and there is no way to limit the rate at which jobs are submitted.

**Comment by pditommaso** [comment]
You can use the nextflow [queueSize](https://www.nextflow.io/docs/latest/config.html#scope-executor) setting or the command line option `-qs` to limit the number jobs submitted to your grid scheduler

**Comment by murphycj** [comment]
Indeed I do use that option.

**Comment by andreas-wilm** [comment]
Hi Paolo, can I vote to open this again? Reason is that we could use that feature as well. It's a workaround for overloaded schedulers. Limiting the queueSize doesn't help there, but throttling the submission frequency. On one system we use, submission can fail if you submit multiple jobs per second and the system is under load, so we throttle the submission rate (in snakemake that is). 
Thanks, Andreas

**Comment by pditommaso** [comment]
It should be doable. Most of the code is related to parse the rate limit format .. ;)

**Comment by pditommaso** [comment]
OK, in the config file you can add the following setting 

```
executor {
  submitRateLimit = '<limit>'
}
```

Where limit is a string representing the max tasks per second, or can follow this syntax:

```
    '1'             // 1 per second
    '5'             // 5 per second
    '100 min'       // 100 per minute
    '100 / 1 s'     // 100 per second
    '100 / 2 s'     // 100 per 2 seconds
    '600 / 5'       // 600 per 5 seconds
    '600 / 5min'    // 600 per 5 minutes
```

Please give it a try defining the var `NXF_VER=0.30.0-SNAPSHOT` eg

```
NXF_VER=0.30.0-SNAPSHOT nextflow run .. etc 
```


I think I'e half-earned  invitation to Singapore ;)

**Comment by andreas-wilm** [comment]
Thanks! What love to have you over for a visit. Need to find some money. Fixing https://github.com/nextflow-io/nextflow/issues/648 might save us some money, which we could use for sponsoring a trip ;)

**Comment by pditommaso** [comment]
;)

**Comment by pditommaso** [comment]
@andreas-wilm have you have chance to test this?

**Comment by andreas-wilm** [comment]
Will get a chance to test this at scale in the next two weeks

**Comment by pditommaso** [comment]
OK, hopefully I will release the version before. I assume that this is fine. If there's any problem open a new issue.

**Comment by RoganGrant** [comment]
Can confirm that limiting job submission rate prevents timing out. 21.0.4 was sending an average of 6 jobs per second, which was too much for our HPC with current traffic. Adding `executor.setRateLimit = '1sec'` to my nextflow config did the trick.
